### PR TITLE
[feat](Nereids): compute func deps in logical plan

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes

This pull request introduces several improvements to the handling of function dependencies and resolves a bug in the LogicalUnion Node:

1. Introduce a direct graph data structure to efficiently store and manage function dependencies. This enhancement allows for better organization and faster access to the dependencies between functions.

2. Implement the computation of function dependencies within each plan. By analyzing the relationships and dependencies among functions in each plan, we can optimize the execution order and identify potential optimizations.

3. Extend the function dependency handling by utilizing unique slots and uniform slots. This extension enables more precise tracking of dependencies, taking into account the specific characteristics of each slot type. It improves the accuracy and reliability of the dependency analysis.

4. Fix a bug discovered in the LogicalUnion Node. The issue has been identified and resolved, ensuring the correct functionality of the node in the presence of function dependencies.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FuncDepsDG.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+/**
+ * Represents a direct graph where each node corresponds to a set of slots in a table,
+ * and each directed edge denotes a dependency where the child node depends on its parent node.
+ * e.g.
+ * if we have a = b, and we can get b -> a and b -> a. The DG is as follows:
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/FuncDepsDG.java:38]: should be?
```suggestion
 * if we have a = b, and we can get b -> a and a -> b. The DG is as follows:
```

**Review by morrySnow — APPROVED** [review]

**Review by jackwener — APPROVED** [review]

**Pull request merged**